From dc2d0e04f61d95e9ab5fa90d14f348982f6290a5 Mon Sep 17 00:00:00 2001
From: Ariel Elior <ariele@broadcom.com>
Date: Wed, 23 Jan 2013 03:21:54 +0000
Subject: [PATCH 600/858] bnx2x: SR-IOV version compatibility bugfix

When posting a message on the bulletin board, the PF calculates crc
over the message and places the result in the message. When the VF
samples the Bulletin Board it copies the message aside and validates
this crc. The length of the message is crucial here and must be the
same in both parties. Since the PF is running in the Hypervisor and
the VF is running in a Vm, they can possibly be of different versions.
As the Bulletin Board is designed to grow forward in future versions,
in the VF the length must not be the size of the message structure
but instead it should be a field in the message itself.

Signed-off-by: Ariel Elior <ariele@broadcom.com>
Signed-off-by: Yuval Mintz <yuvalmin@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 4c133c39ae1bc541b3db5903a680da738a0c57e3)

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c | 2 +-
 drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.c  | 4 ++--
 drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.h  | 3 ++-
 3 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
index 08db503..b681444 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
@@ -3124,7 +3124,7 @@ u32 bnx2x_crc_vf_bulletin(struct bnx2x *bp,
 {
  return crc32(BULLETIN_CRC_SEED,
    ((u8 *)bulletin) + sizeof(bulletin->crc),
-   BULLETIN_CONTENT_SIZE - sizeof(bulletin->crc));
+   bulletin->length - sizeof(bulletin->crc));
 }
 
 /* Check for new posts on the bulletin board */
diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.c
index 850aad3..1cbc04c 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.c
@@ -1631,7 +1631,6 @@ int bnx2x_post_vf_bulletin(struct bnx2x *bp, int vf)
  dma_addr_t pf_addr = BP_VF_BULLETIN_DMA(bp)->mapping +
   vf * BULLETIN_CONTENT_SIZE;
  dma_addr_t vf_addr = bnx2x_vf(bp, vf, bulletin_map);
- u32 len = BULLETIN_CONTENT_SIZE;
  int rc;
 
  /* can only update vf after init took place */
@@ -1641,11 +1640,12 @@ int bnx2x_post_vf_bulletin(struct bnx2x *bp, int vf)
 
  /* increment bulletin board version and compute crc */
  bulletin->version++;
+ bulletin->length = BULLETIN_CONTENT_SIZE;
  bulletin->crc = bnx2x_crc_vf_bulletin(bp, bulletin);
 
  /* propagate bulletin board via dmae to vm memory */
  rc = bnx2x_copy32_vf_dmae(bp, false, pf_addr,
       bnx2x_vf(bp, vf, abs_vfid), U64_HI(vf_addr),
-      U64_LO(vf_addr), len/4);
+      U64_LO(vf_addr), bulletin->length / 4);
  return rc;
 }
diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.h b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.h
index 21d2985..f5dbb3c 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.h
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_vfpf.h
@@ -318,7 +318,8 @@ struct pf_vf_bulletin_content {
  u32 crc;   /* crc of structure to ensure is not in
       * mid-update
       */
- u32 version;
+ u16 version;
+ u16 length;
 
  aligned_u64 valid_bitmap; /* bitmap indicating which fields
       * hold valid values
-- 
1.8.3.1